drivers: spi: spi_nrfx_spim: Add support for RX buffer from RAM region

awojasinski opened the pull request:
This patch adds support for RX buffer placed by a linker in memory region defined in SPIM devicetree node. The buffer is placed in memory region defined as devicetree node. The memory region node's reference is then stored in `memory-regions` property of SPIM node.

Added build time assertion to check if `CONFIG_SPI_NRFX_RAM_BUFFER_SIZE` Kconfig symbol has value greater than 0 when given SPIM node has `memory-region` property.

Commit:
drivers: spi: spi_nrfx_spim: Add support for RX buffer from RAM region

This patch adds support for RX buffer placed by a linker in
memory region defined in SPIM devicetree node. The buffer is placed
in memory region defined as devicetree node. The memory region node's
reference is then stored in `memory-regions` property of SPIM node.

Added build time assertion to check if `CONFIG_SPI_NRFX_RAM_BUFFER_SIZE`
Kconfig symbol has value greater than 0 when given SPIM node has
`memory-region` property.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

awojasinski-nordicsemi commented:
cc @anangl

MaureenHelm commented:
@anangl please take a look

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -627,15 +652,18 @@ static int spi_nrfx_init(const struct device *dev)
 		.irq_connect = irq_connect##idx,			       \
 		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(SPIM(idx)),		       \
-		.max_chunk_len = BIT_MASK(SPIM_PROP(idx, easydma_maxcnt_bits)),\
+		.max_chunk_len = SPI_NRFX_SPIM_MAX_CHUNK_LEN(idx),	       \
```
This should not be changed. When the buffer supplied in a request is DMA-accessible, then the internal buffer is not used for the transaction and its size should not limit the chunk length. Thus, `max_chunk_len` should be based on the hardware limit as it was so far.
For the case when the internal buffer is used, the chunk length should be limited at runtime by
```
if (chunk_len > CONFIG_SPI_NRFX_RAM_BUFFER_SIZE) {
	chunk_len = CONFIG_SPI_NRFX_RAM_BUFFER_SIZE;
}
```

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -310,25 +311,34 @@ static void transfer_next_chunk(const struct device *dev)
 		}
 
+#ifdef SPI_BUFFER_IN_RAM
+		if (chunk_len > CONFIG_SPI_NRFX_RAM_BUFFER_SIZE) {
+			chunk_len = CONFIG_SPI_NRFX_RAM_BUFFER_SIZE;
+		}
```
This should not be applied always, but only when a given transfer actually uses the internal buffer.

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -599,15 +618,19 @@ static int spi_nrfx_init(const struct device *dev)
+			[CONFIG_SPI_NRFX_RAM_BUFFER_SIZE]		       \
+			SPIM_MEMORY_SECTION(idx);			       \
+		 static uint8_t spim_##idx##_rx_buffer			       \
 			[CONFIG_SPI_NRFX_RAM_BUFFER_SIZE]		       \
```
Since the `SPI_NRFX_RAM_BUFFER_SIZE` option applies now to both TX and RX, it would be good to also update its description as it still mentions transmitting from flash only.

Commit:
drivers: spi: nrfx: Update doc for RAM_BUFFER_SIZE Kconfig symbol

Update documentation for SPI_NRFX_RAM_BUFFER_SIZE Kconfig symbol
to reflect new usage of it. Now the symbol specifies size of RX buffer.
The change introducing support for RX buffer placed by a linker in
memory region defined in SPIM devicetree node is in a parent commit
of that one.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -310,25 +311,35 @@ static void transfer_next_chunk(const struct device *dev)
+		if (spi_context_tx_buf_on(ctx) &&
+		    !nrf_dma_accessible_check(&dev_config->spim.p_reg, tx_buf)) {
+
 			if (chunk_len > CONFIG_SPI_NRFX_RAM_BUFFER_SIZE) {
 				chunk_len = CONFIG_SPI_NRFX_RAM_BUFFER_SIZE;
 			}
```
This should be also done in the RX path. There may be an RX-only transfer to a buffer not accessible by DMA.

anangl commented on drivers/spi/Kconfig.nrfx:
```diff
@@ -59,12 +59,13 @@ config SPI_NRFX_RAM_BUFFER_SIZE
+	  unreachable by EasyDMA. The size is specified in bytes. A size of 0 means that
+	  this feature should be disabled, and the application must then take care of not
 	  supplying buffers located in flash to the driver, otherwise such
 	  transfers will fail.
```
This part still mentions buffers located in flash, so it should be updated.

anangl commented on drivers/spi/Kconfig.nrfx:
```diff
@@ -59,14 +59,15 @@ config SPI_NRFX_RAM_BUFFER_SIZE
-	  should be disabled, and the application must then take care of not
-	  supplying buffers located in flash to the driver, otherwise such
-	  transfers will fail.
+	  SPIM peripherals cannot transmit data directly from certain memory regions or receive
+	  data to any memory region. Therefore, buffers in RAM need to be provided
+	  for each instance of SPI driver using SPIM peripheral, so that the driver can copy
+	  to one of them a chunk of data - from memory unreachable by EasyDMA - to transmit it.
+	  The second buffer is used to receive data when the user's buffer is
+	  unreachable by EasyDMA. The size is specified in bytes. A size of 0 means that
+	  this feature should be disabled, and the application must then take care of not
+	  supplying buffers located in a memory region unaccessible by EasyDMA to the driver,
+	  otherwise such transfers will fail.
```
How about using the following:
```
	  Because of using EasyDMA, SPIM peripherals cannot use transmit and
	  receive buffers from all memory locations. They are restricted to
	  buffers located in certain RAM memories only. Therefore, each SPIM
	  driver instance needs to use an intermediate local RAM buffer,
	  to transfer data in chunks not exceeding the size of that buffer,
	  and to copy those chunks between the local buffer and the one
	  specified in the transfer request if the latter is not accessible
	  by EasyDMA.

	  This option specifies the size in bytes of such local RAM buffers
	  for both TX and RX paths. A size of 0 means that this feature should
	  be disabled and the driver user must take care of not making transfer
	  requests with buffers not accessible by EasyDMA since such transfers
	  will fail.
```

awojasinski-nordicsemi commented:
@tbursztyka @teburd please review the PR

nordic-krch commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -639,7 +666,10 @@ static int spi_nrfx_init(const struct device *dev)
 	};								       \
-	BUILD_ASSERT(!DT_NODE_HAS_PROP(SPIM(idx), wake_gpios) ||	       \
+	BUILD_ASSERT(!SPIM_HAS_PROP(idx, memory_regions) ||		       \
+		     CONFIG_SPI_NRFX_RAM_BUFFER_SIZE != 0,		       \
```
is it required that this option is enabled when `memory_regions` exists? Isn't there an option for user to provide buffers from allowed location?

anangl replied:
Good point. This is an unnecessary limitation. We should get rid of this assertion.

awojasinski-nordicsemi replied:
I thought that it's worth checking but I don't have a strong opinion on that. Removed the check
